Translation banner logic for non-markdown pages

wackerow opened the pull request:
## Description
The translation banner indicating that content is not yet translated is handled with logic from the page component. This works already on markdown pages, as this logic is included in `[...slug]`, but it was not included for any of the recent React page components that have been added.

- Adds a `existsNamespace ` utility function to `translations.ts` which checks if the `.json` file for the namespace requested exists for that locale. Returns `true` if translated content exists, else false.
- Implements this function inside `getStaticProps` for ALL pages.

## Related Issue
https://www.notion.so/efdn/NextJS-general-QA-session-Dec-14th-cdec3085d96c4608a96b80772755f53a?pvs=4#c688e24c1d6547b69a43e8a40978515d

## Note
Logic does NOT include "outdated" content (those with translations available, but not fully up-to-date with English), only when translated content is not available.

wackerow committed:
add getIsContentNotTranslated utility

wackerow committed:
add contentNotTranslated prop to gas page

wackerow committed:
comment function

pettinarip commented on src/lib/utils/translations.ts:
```diff
@@ -210,3 +212,19 @@ const getRequiredNamespacesForLayout = (layout?: string) => {
+ * @param existsSync Passed from fs library
+ * @returns false if the primary namespace .json file exists, true otherwise
+ */
+export const getIsContentNotTranslated = (
```
I'd prefer a positive naming always. Improves the reading experience for devs. Then you can do the negative on the consumer's side

```ts
const isContentNotTranslated = !existsNamespace(...)

// or even better
const isContentTranslated = existsNamespace(...)

if (!isContentTranslated) {
  ...
}
```

Reading this function name without any context makes it difficult to understand its purpose.

And we're returning a boolean, no need the `get` prefix on the name.

```suggestion
export const existsNamespace = (
```

wackerow replied:
Agree... had named to match the existing boolean name we were using (`contentNotTranslateds`), but can update this

pettinarip commented on src/lib/utils/translations.ts:
```diff
@@ -210,3 +212,19 @@ const getRequiredNamespacesForLayout = (layout?: string) => {
+export const getIsContentNotTranslated = (
+  locale: string,
+  primaryNs: string,
+  existsSync: (arg0: string) => boolean
```
This is implementation detail. No need to receive this function as an argument.

Consumer only cares about `locale` and `namespace`. The function should take care of the rest.

```suggestion
```

wackerow replied:
Agree... Had done it this way initially to avoid importing from `fs` inside the `translations.ts` file, which is consumed by other non-page components that won't have access to `fs`:

<img width="611" alt="image" src="https://github.com/ethereum/ethereum-org-fork/assets/54227730/878a3f16-85a8-485e-9b8e-114101bbae61">

I think switching this to a dedicated util file would be the best way to go. That will avoid this conflict, and avoids passing a prop in this manner.

pettinarip reviewed (changes requested):
Nice addition @wackerow. Left comments about the implementation.

wackerow committed:
mv existsNamespace to own util file

requires fs import, which conflicts if placed inside `translations.ts` when that util is imported from components where filesystem is not available. Moved to isolated util file. Renamed for "positive" naming

pettinarip commented on src/lib/utils/existsNamespace.ts:
```diff
@@ -0,0 +1,16 @@
+  namespace: string,
+): boolean => {
+  const nsJsonPathForLocale = join("../intl", locale, namespace + ".json")
+  return !existsSync(nsJsonPathForLocale)
```
Sorry, should have detailed better, this should be positive now.
```suggestion
  return existsSync(nsJsonPathForLocale)
```

wackerow replied:
Oops, my bad, meant to get that, good catch

pettinarip committed:
Update src/lib/utils/existsNamespace.ts

wackerow commented:
Have updated per suggestions. Pulled `existsNamespace` into a new util file to isolate this function which utilized `fs`.

Welcome more comments on this approach, but will place in draft mode for now while we await completion of remaining pages. Will push remaining page-component updates if we're okay with this approach, and when all pages are ready we can merge.

wackerow committed:
update types/interfaces and RootLayout

remove redundant namespace for learning-tools from translations

wackerow committed:
implement contentNotTranslated page prop

all available pages; uses consistent formatting for getStaticProps throughout with new shared BasePageProps type

wackerow committed:
Merge branch 'dev' into translation-banner-logic

wackerow committed:
fix: add contentNotTranslated prop to stablecoins

wackerow committed:
fix: add missing import

wackerow committed:
fix: add contentNotTranslated prop to layer-2

wackerow committed:
fix: add contentNotTranslated prop to wallet pages

wackerow committed:
fix: add contentNotTranslated prop to what-is-ethereum

wackerow committed:
fix: add contentNotTranslated prop to acknowledgements

wackerow committed:
fix: add contentNotTranslated prop to contributors

wackerow committed:
refactor: use BasePageProps for stablecoins

wackerow committed:
Merge branch 'dev' into translation-banner-logic

wackerow committed:
Merge branch 'dev' into translation-banner-logic

pettinarip commented on src/pages/developers/learning-tools.tsx:
```diff
@@ -115,23 +115,25 @@ const StackContainer = (props: ChildOnlyProp) => (
     "/developers/learning-tools"
   )
+
+  const contentNotTranslated = !existsNamespace(locale!, requiredNamespaces[2])
```
Would love to have a more deterministic way to get the specific namespace in future iterations.

pettinarip commented on src/pages/quizzes.tsx:
```diff
@@ -35,6 +36,22 @@ const handleGHAdd = () =>
     eventName: "GH_add",
   })
 
+export const getStaticProps = (async ({ locale }) => {
```
Nice, thanks for reorder this.

pettinarip commented on src/lib/types.ts:
```diff
@@ -29,6 +29,15 @@ export type AppPropsWithLayout = AppProps & {
+  lastDeployDate: string
+}
+
+export type BasePageProps = SSRConfig & Pick<Root, "contentNotTranslated" | "lastDeployDate">
```
👍🏼

pettinarip reviewed (approved):
Looks good @wackerow. Nice refactors as well.
